### ISSUE TITLE
Upgrade dpd to work with version 0.39

### ISSUE THREAD

**Issue by GibbsConsulting** [GibbsConsulting/django-plotly-dash#127]
The new version of plotly dash (0.39) does not work with dpd - components do not render and some adjustment to the dash template is needed.

**Comment by NeodarZ** [comment]
Hi, only 0.39 ? 
Why not directly move to the last version 0.44 ?

**Comment by eddy-ojb** [comment]
@NeodarZ, according to the official changelog and documentation, the current stable version is 0.39:

https://github.com/plotly/dash/blob/master/CHANGELOG.md
https://dash.plot.ly/installation

**Comment by NeodarZ** [comment]
Arf, yes sorry for my mistake, I confused dash and dash-core-components... :/

**Comment by ivan-katkov** [comment]
This is highly desirable 😍 improvement because Multiple Outputs have been released 🎉 in the 0.39.0 version (see [here](https://community.plot.ly/t/multiple-outputs-in-dash-now-available/19437))

**Comment by NinoZombie** [comment]
Thank you for your great lib, @GibbsConsulting ! :)
There is already released new (0.40.0) Dash. Do you have any news about "some adjustment to the dash template" ? I tried myself to figure out what adjustments need to be made, but I could not :(

**Comment by vantaka2** [comment]
If someone can break this down to individual tasks, I'm more then happy to help out where I can.

**Comment by jgraham0331** [comment]
Is there anything I can do to help out with this?  I'd like to switch to the new version of dash but I rely too heavily on dpd to upgrade.  I haven't tested 0.39 enough to know what's broken.

**Comment by GibbsConsulting** [comment]
Not had a chance to investigate much yet, but I do think that there are at least two chunks of work:

- handle multiple return values from a callback
- something has changed in the way apps are rendered - there is now a `{%renderer%}` tag in the Dash template - so dpd has to also reflect this change

There could well be more. I'm also hopeful that these can be dealt with separately.

For reference, the dash changeset is [here](https://github.com/plotly/dash/compare/v0.38.0...0.39.0-issue591)

**Comment by sdementen** [comment]
Maybe this https://github.com/plotly/dash/pull/436 is more specific re the change for the multi-output

**Comment by GibbsConsulting** [comment]
@sdementen thanks for the useful link

**Comment by sdementen** [comment]
@GibbsConsulting maybe worth to ask the dash team (or the members behind the pull request) what they have changed to understand what needs to be changed for dpd (just high level view). Do they know about dpd?

**Comment by jgraham0331** [comment]
dpd is referenced quite a bit on the Dash forums but I don't think Django is part of the core scope of the project.  I know for what I do, if it wasn't for dpd I'd have abandoned Dash and gone with Bokeh.  Our entire app runs on Django and I'm not switching to Flask just to get Dash working.  I've seen the folks at Dash offer to include support for Django if someone would fund the project to do so.

**Comment by jgraham0331** [comment]
> Not had a chance to investigate much yet, but I do think that there are at least two chunks of work:
> 
> * handle multiple return values from a callback
> * something has changed in the way apps are rendered - there is now a `{%renderer%}` tag in the Dash template - so dpd has to also reflect this change
> 
> There could well be more. I'm also hopeful that these can be dealt with separately.
> 
> For reference, the dash changeset is [here](https://github.com/plotly/dash/compare/v0.38.0...0.39.0-issue591)

@GibbsConsulting : I am going to look into both issues to the greatest extent possible.  I don't have the understanding of your code or Dash's to make much of an impact here but if I come up with anything I will let you know.

**Comment by jchang10** [comment]
https://github.com/GibbsConsulting/django-plotly-dash/compare/master...jchang10:jchang10-patch-1

I just took a look at it. To support our existing apps, it was a fairly minimal set of changes above. This does not support multiple outputs. Single output seems to work with all the latest revisions.

The changes were minimal. However, a few days ago, a set of merge conflicts occurred. Unfortunately, I guess I am not the best at github, so I just managed to get the above link to work.

Hopefully, somehow can get a better handle of things than myself. The things to sort out:
1. Multiple output support. Only single outputs working in my changes.
2. The slew of DPD changes were related to local assets. Somehow they need to get merged in too.
3. I noticed a few other DASH changes, related to callback context's, for example, that I did not bother keeping either.
